console_backend/src/output.rs

```diff
@@ -50,42 +50,59 @@ impl CsvLogging {
 pub type SbpLogging = cc::SbpLogging;
 
 #[derive(Debug)]
-pub enum SbpLogger {
-    Sbp(File),
-    Json(JsonEncoder<File, CompactFormatter>),
+pub struct SbpLogger<W> {
+    logger: SbpLogging,
+    out: W,
+    path: Option<PathBuf>,
 }
 
-impl SbpLogger {
-    pub fn new_sbp<P: AsRef<Path>>(filepath: P) -> Result<SbpLogger> {
-        Ok(SbpLogger::Sbp(File::create(filepath)?))
+impl<W: Write> SbpLogger<W> {
+    pub fn new(logger: SbpLogging, out: W, path: Option<PathBuf>) -> Self {
+        Self { logger, out, path }
     }
```

Comment:
I don't love this api design where the caller to the constructor has to path both writer W & Optional PathBuf, and it is up to the caller to know if `PathBuf` is some, we will require that file to exist whenever `serialize` is called.

what's to stop a person from calling something that looks like
```rust 
    let f = File::create("foo.txt")?;
    let mut f = BufWriter::new(f);
    let logger = SbpLogger::new(SbpLogging::SBP, f, None);
}
```
I don't really have a great suggestion here, but I think it may be better to design it as follows:

```rust
pub enum LoggingOutput<W: Write> {
    File { out: W, path: PathBuf } 
}

#[derive(Debug)]
pub struct SbpLogger<W: Write> {
    logger: SbpLogging,
    out: W,
    path: Option<PathBuf>,
}

impl<W: Write> SbpLogger<W> {
    pub fn new(logger: SbpLogging,  output: LoggingOutput) -> Self {
        match output {
            LoggingOutput:File {out, path} => Self { logger, out, path: Some(path} },
        }
    }
```

You might argue YAGNI, with the enum of one type. However, I think if you aren't going to need an enum here, I don't understand why you would need the new fn to take a generic writer & an optional pathbuf.

Reply:
hmm yeah, i wasn't sure if abstraction is even needed here since idk if we would ever want to emit this data anywhere else - uploading ? or stdout or something? (which might also introduce some funky logic as to how we want to reflect # of bytes written or check the success of those) so kinda went in between

this just kinda explicitly state a link for the write to the pathbufs to address the UI not being updated (oOoOo spooky inodes 👻)

but if we do, maybe just have a callback for specific logger types would be best instead of a pathbuf to allow more flexibility 🤔

hmm maybe file lock would be nice or 
maybe best to just keep this simply as sbp file logger with pathbuf without doing funky `impl Write` stuff

Reply:
that's fine but then I would also remove the option on the `PathBuf`